contents/java/JavaReflections.md

```diff
@@ -0,0 +1,291 @@
+itself. This is very useful if you are interested to implement tests cases to test the class.
+
+### The Basics of Reflections
+
+Before getting started with reflections in Java, it is important to realize that classes themselves is an object. Every
+unique `Object` is assigned an immutable `Class` object by the JVM. This immutable `Class` object is fundamentally 
+different from *instances* of a class. The class object itself holds information about its name, the package it resides 
+in and other important information while an instance of a class holds the instanced values and methods as defined
+in the class.
+
+Take for example the following class:
+
+```java
+public class Student {
+
+	private final String name;
+	private final String gender;
+
+	public Student(String name, String gender) {
+		this.name = name;
+		this.gender = gender;
+	}
+
+    // Other methods here...
+}
+```
+
+An instance of a class can be created as usual using the `new` keyword
+
+```java
+Student john = new Student("John Doe", "Male");
+```
+
+However, it is also possible to get information about the class itself because the class itself is an `Object` like so:
+
+```java
+Class<Student> studentClass = Student.class;
+```
+
+It also means that you can store any `Class` object in any data structure for future retrieval. This is the main entry
+point for Java's reflections. You can now get the name of the class, create new instances of the class, observe its
+public/private fields - the possibilities are endless!
+
+### Getting Started
+
+There are many webpages dedicated to explaining the details of reflections in Java; so this will not repeat what is
+being made readily available on the web. One good place to start is [http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html](http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html)
+
+What this page will do instead is to give some examples to demonstrate what reflections can do and give you some
+ideas on how you can use reflections in your project.
+
+One important point is that while Java reflections are powerful, its implementations are anything but. There are however
+some libraries out there to make life easier such as the [Google's Guava
+library](https://github.com/google/guava/wiki/ReflectionExplained) which contains many utility methods and classes to
+make your life easier.
+
+### Applications
+
+#### Accessing private fields
+
+A good example of reflections is to get a private field of another class. While this should
+optimally be solved by modifying the field visibility to `protected` or `public`, sometimes it is not possible to do so
+becuase you might not have any access to the code (for example codes in libraries or frameworks).
+
+For the sake of simplicity, let us use the previous example of a `Animal`. The class definition can look like this.
```

Comment:
"let us use the previous example of a  Animal". Was the definition of the animal class moved down?